src/cpp/src/lora_adapter.cpp

```diff
@@ -0,0 +1,1296 @@
+#include "openvino/op/convert.hpp"
+#include "openvino/op/convolution.hpp"
+#include "openvino/op/matmul.hpp"
+#include "openvino/op/concat.hpp"
+#include "openvino/op/reshape.hpp"
+#include "openvino/op/broadcast.hpp"
+#include "openvino/op/read_value.hpp"
+#include "openvino/op/assign.hpp"
+#include "openvino/op/transpose.hpp"
+#include "openvino/op/util/variable.hpp"
+#include "openvino/pass/pattern/matcher.hpp"
+#include "openvino/pass/pattern/op/wrap_type.hpp"
+#include "openvino/pass/graph_rewrite.hpp"
+#include "openvino/pass/manager.hpp"
+#include "openvino/runtime/core.hpp"
+
+#include "openvino/genai/lora_adapter.hpp"
+
+extern "C" {
+    #include "safetensors.h"
+}
+
+// If set to 1, the empty tensors will be used to switch LoRA adapter off.
+// FIXME: Fix the plugins and set to 1 permanently.
+#define EMPTY_TENSORS_SUPPORTED_IN_MATMUL 0
+
+// If set to 1, LoRA state tensors will have the original type of LoRA adapter come from safetensors file.
+// If there are multiple LoRA adapters are applied, then negotiation between them happens.
+// If set to 0, LoRA state etnsors are always have type f32.
+// FIXME: Fix the plugins and set to 1 permanently.
+#define FP16_BF16_TENSORS_SUPPORTED_IN_STATE 0
+
+// FIXME: Remove or move to a dedicated common header
+#ifdef NDEBUG
+    #define DEBUG_PRINT(X) do {} while(false)
+#else
+    #define DEBUG_PRINT(X) do { std::cerr << "[ DEBUG ] " << X << "\n"; } while(false)
+#endif
+
+namespace {
+
+using NodePtr = std::shared_ptr<ov::Node>;
+using ov::NodeVector;
+using namespace ov::op;
+
+// FIXME: Use ov::AlignedBuffer instead of std::vector. ov::AlignedBuffer is not available in public OV API
+using Buffer = std::vector<char>;
+using BufferPtr = std::shared_ptr<Buffer>;
+using ConstantVector = std::vector<std::shared_ptr<v0::Constant>>;
+
+
+// Holds usual LoRA parameters alpha, A and B of a given type.
+template <typename T>
+struct LoRAParts {
+    T alpha, A, B;
+
+    LoRAParts() = default;
+    LoRAParts(const T& alpha, const T& A, const T& B) : alpha(alpha), A(A), B(B) {}
+
+    template <typename Other>
+    LoRAParts(const LoRAParts<Other>& other) : alpha(other.alpha), A(other.A), B(other.B) {}
+};
+
+
+using LoRAWeight = LoRAParts<std::shared_ptr<v0::Constant>>;
+using LoRANode = LoRAParts<std::shared_ptr<ov::Node>>;
+using LoRAPartsParser = LoRAParts<std::function<std::optional<std::string>(const std::string& name)>>;
+using LoRATensors = std::map<std::string, LoRAWeight>;
+
+
+// Read binary file to memory.
+BufferPtr read_file_helper(const std::string& filename) {
+    std::ifstream file(filename, std::ios::binary | std::ios::ate);
+    OPENVINO_ASSERT(file.is_open(), "Cannot open file with LoRA weights: ", filename);
+
+    size_t filesize = file.tellg();
+    auto buffer = std::make_shared<Buffer>(filesize);
+    file.seekg(0, std::ios::beg);
+    // TODO: Use mmapped AlignedBuffer as ov::Core::read_model can do, necessary functionality is not available in public OV APIю
```

Comment:
it's available in Developer package, which is used in Jenkins to build genAI.
You can add conditional code which uses `AlignedBuffer` depending on whether are building against Developer package or Public API

Reply:
Added explicit TODO into the PR list. Will be fixed in the future PRs.